### DISCUSSION TITLE
How to adaptively match the docker client sdk version

### DISCUSSION THREAD

**Discussion by moluzhui** [moby/moby#45381] [Q&A]
I have a project that needs to manage containers through **docker sdk**, and I use [WithAPIVersionNegotiation](https://stackoverflow.com/a/57260324/11264544) to reduce the version.

The example is as follows
```
cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
	ctx := context.Background()
	if err != nil {
		panic(err)
	}
	defer cli.Close()

	cmdStatementExecuteScript := cmdStatement
	optionsCreateExecuteScript := types.ExecConfig{
		AttachStdout: true,
		AttachStderr: true,
		Cmd:          cmdStatementExecuteScript,
	}

	response, err := cli.ContainerExecCreate(ctx, containerID, optionsCreateExecuteScript)
	if err != nil {
		panic(err)
	}
```
But when I deploy the service on multiple machines, the following error occurs because of the *different docker versions on each machine*.
```
panic: Error response from daemon: client is newer than server (client API version: 1.24, server API version: 1.22)
```
> docker version
```
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:39:25 2016
 OS/Arch:      linux/amd64
```
`WithAPIVersionNegotiation` config doesn't seem to be downgraded to the **required version**.

Is there some way to satisfy adaptive all versions by coding? Looking forward to your reply.

> PS: Use `WithAPIVersionNegotiation` to **take effect** when *Error response from daemon: client version 1.42 is too new. Maximum supported API version is 1.39* appears.

go.mod:
```
	github.com/docker/distribution v2.8.1+incompatible // indirect
	github.com/docker/docker v23.0.4+incompatible
	github.com/docker/go-connections v0.4.0 // indirect
	github.com/docker/go-units v0.5.0 // indirect
```

**Comment by thaJeztah** [comment]
Version Negotiation was implemented in Docker 1.13, and the client won't negotiate automatically to API versions older than that.

It looks like you're running a really old version of docker (1.10.3) that was released more than 7 Years ago https://github.com/moby/moby/releases/tag/v1.10.3

That version reached EOL a long time ago, and has many known vulnerabilities; I'd _highly_ recommend upgrading to a currently maintained version.
